addons/sale/models/sale_order_line.py

```diff
@@ -890,7 +889,13 @@ def _compute_analytic_distribution(self):
     @api.depends('product_id', 'state', 'qty_invoiced', 'qty_delivered')
     def _compute_product_updatable(self):
         for line in self:
-            if line.state in ['done', 'cancel'] or (line.state == 'sale' and (line.qty_invoiced > 0 or line.qty_delivered > 0)):
+            if line.state == 'cancel':
+                line.product_updatable = False
+            elif line.state == 'sale' and (
+                line.order_id.locked
+                or line.qty_invoiced > 0
+                or line.qty_delivered > 0
+            ):
```

Comment:
I can see why it's like this, but I'd prefer something like:

```suggestion
            is_locked = line.order_id.locked
            delivered_or_invoiced = line.qty_invoiced > 0 or line.qty_delivered > 0
            not_updatable_sale = line.state == 'sale' and (is_locked or delivered_or_invoiced)
            line.product_updatable = not(line.state == 'cancel' or not_updatable_sale)
```